### DISCUSSION TITLE
Is there any function in pyright implementation to reconvert the parse tree into python source code?

### DISCUSSION THREAD

**Discussion by Andres981231** [microsoft/pyright#2661] [Q&A] [answered]
As title. Pyright convert the python source code into parse tree before further analysis, so I wonder whether I can modify the parse tree and reconvert it into the python source code with the current pyright implementation? Thanks a lot.

**Comment by erictraut** [comment, marked as answer]
No, the parse tree that pyright builds does not contain enough information to reclaim the original text. It is not intended for that purpose. Only information required for type checking and language server features are included in the parse tree.

**Reply by erictraut** [reply]
There is a method called `printExpression` in `parseTreeUtils.ts` that can be used to print a text representation of an expression node in the parse tree, but it doesn't make any attempts to retain the original text form. It's intended to be used for some error and logging messages.

**Reply by Andres981231** [reply]
Thanks!

**Comment by Josverl** [comment]
@Andres981231: not pyright related, but possibly [LibCST](https://libcst.readthedocs.io/en/latest/why_libcst.html#libcst)  might be of interest to you for code mods.

**Reply by Andres981231** [reply]
Thanks for your sharing.
